lib/change_stream.js

```diff
@@ -215,13 +241,93 @@ class ChangeStream extends EventEmitter {
   }
 }
 
-// Create a new change stream cursor based on self's configuration
-var createChangeStreamCursor = function(self) {
-  if (self.resumeToken) {
-    self.options.resumeAfter = self.resumeToken;
+class ChangeStreamCursor extends Cursor {
+  constructor(bson, ns, cmd, options, topology, topologyOptions) {
+    // TODO: spread will help a lot here
+    super(bson, ns, cmd, options, topology, topologyOptions);
+
+    options = options || {};
+    this._resumeToken = null;
+    this.startAtOperationTime = options.startAtOperationTime;
+
+    if (options.startAfter) {
+      this.resumeToken = options.startAfter;
+    } else if (options.resumeAfter) {
+      this.resumeToken = options.resumeAfter;
+    }
+  }
+
+  set resumeToken(token) {
+    this._resumeToken = token;
+    this.emit('resumeTokenChanged', token);
+  }
+
+  get resumeToken() {
+    return this._resumeToken;
+  }
+
+  get resumeOptions() {
+    if (this.resumeToken) {
+      return { resumeAfter: this.resumeToken };
+    }
+
+    if (this.startAtOperationTime && maxWireVersion(this.server) >= 7) {
+      return { startAtOperationTime: this.startAtOperationTime };
+    }
+
+    return null;
+  }
+
+  _initializeCursor(callback) {
+    super._initializeCursor((err, result) => {
+      if (err) {
+        callback(err, null);
+        return;
+      }
+
+      const response = result.documents[0];
+
+      if (
+        this.startAtOperationTime == null &&
+        this.resumeAfter == null &&
+        this.startAfter == null &&
+        maxWireVersion(this.server) >= 7
+      ) {
+        this.startAtOperationTime = response.operationTime;
+      }
+
+      const cursor = response.cursor;
+      if (cursor.firstBatch.length === 0 && cursor.postBatchResumeToken) {
+        this.resumeToken = cursor.postBatchResumeToken;
+      }
+
+      this.emit('response');
```

Comment:
is the response event documented anywhere?

Reply:
I didn't want to document it, because I don't really want anybody depending on it. At very least, if we have a documented event the event should include the actual response..

Reply:
Can we add a doclet along the lines of:

```js
/**
 * @event response
 * internal event DO NOT USE
 * @ignore 
 */
```

Reply:
added